### PR TITLE
Qubicle

### PR DESCRIPTION
**Pull request by mark2600**

### CODE REVIEW

File: applications/blackhole/src/blackhole_socket_callback.erl [hunk 1]
```diff
@@ -55,6 +59,24 @@ recv(_SessionPid, SessionId, Message, Context) ->
     lager:info("receive unknown message ~p on socket ~p", [Message, SessionId]),
     {'ok', Context}.
 
+amqp_send(Context, Data) ->
```
Comment (jamesaimonetti) [inline applications/blackhole/src/blackhole_socket_callback.erl]: This function looks so lonely without a spec!

File: applications/blackhole/src/blackhole_socket_callback.erl [hunk 2]
```diff
@@ -55,6 +59,24 @@ recv(_SessionPid, SessionId, Message, Context) ->
...
+            {Message} = Data,
+            {Keys, _} = lists:unzip(Message),
+
+            SendMessage = Message ++ [
+                {<<"Msg-ID">>, wh_util:rand_hex_binary(16)}
+                | wh_api:default_headers(<<"qubicle">>, <<"1.0">>)
+            ],
+
+            {'ok', Payload} = wh_api:build_message(SendMessage, [], Keys),
+            amqp_util:basic_publish(<<"qubicle">>, <<"qubicle.recipient">>, Payload, ?DEFAULT_CONTENT_TYPE);
```
Comment (jamesaimonetti) [inline applications/blackhole/src/blackhole_socket_callback.erl]: This seems awfully specific to qubicle; maybe is should be part of `bh_qubicle`?

File: applications/callflow/src/module/cf_qubicle.erl [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+join_session_to_queue(Call, SessionId, Queue) ->
+    QueueName = wh_json:get_value(<<"queue_name">>, Queue),
+    lager:info("joining ~p (~p) to queue ~p", [whapps_call:call_id(Call), SessionId, QueueName]),
+
+    Args = wapi_qubicle_session:join_queue_args(Call, SessionId, Queue),
+    cf_exe:send_amqp(Call, Args, fun wapi_qubicle_session:join_queue/1),
+    wait_for_join_response(Call, SessionId, Queue).
+
+wait_for_join_response(Call, SessionId, Queue) ->
+    receive {'amqp_msg', JObj} ->
```
Comment (jamesaimonetti) [inline applications/callflow/src/module/cf_qubicle.erl]: `case whapps_call_command:receive_event(10*?MILLISECONDS_IN_SECOND) of`

File: applications/callflow/src/module/cf_qubicle.erl [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        case wapi_qubicle_session:join_queue_response(JObj) of
+            'invalid' ->
+                wait_for_join_response(Call, SessionId, Queue);
```
Comment (jamesaimonetti) [inline applications/callflow/src/module/cf_qubicle.erl]: Do you want to wait another 10 seconds if the JObj is invalid? Or decrement time spent so far and loop with a smaller timeout?

File: applications/blackhole/src/blackhole_socket_callback.erl (hunk 2 above)
Comment (jamesaimonetti) [inline applications/blackhole/src/blackhole_socket_callback.erl]: Or should maybe lookup based on session id to figure out how to publish? This isn't a generic solution in a generic `recv/4` function.

File: applications/crossbar/priv/couchdb/schemas/qubicle_queues.json [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            "default": "queue_basic",
+            "description": "The type of queue",
+            "enum": [
+                "queue_basic"
+            ],
+            "name": "Type",
+            "required": true,
+            "type": "string"
+        },
+        "tick_time": {
```
Comment (jamesaimonetti) [inline applications/crossbar/priv/couchdb/schemas/qubicle_queues.json]: I have been preferring adding unit suffixes to make it obvious, `tick_time_ms`

**Comment by k-anderson** [comment]
@mark2600 just a heads up I am closing this erronous PR but you should still take a look at @jamesaimonetti 's comments ;)

**Pull request closed without merging**